### PR TITLE
Instance View

### PR DESCRIPTION
**Pull request by Sjmarf**
Re-instates `InstanceView`, with the "About" and "Details" tabs. Quite a bit of the line count is just copy-pasting `InstanceDetailsView` from v1.

You can block/unblock instances. This uses the [new MlemMiddleware `BlockList` class](https://github.com/mlemgroup/MlemMiddleware/pull/31) to keep track of which instances are blocked. In v1, Mlem wouldn't remember whether an instance is blocked or not after you leave the `InstanceView`, but this should keep track of it all the time.

There's also a "visit" action in InstanceView that opens the instance as a temporary guest account. When you switch back to a normal account, the temporary guest account doesn't remain in the list. If you want to keep it, you can long-press on the account in the quick switcher and select "Keep".

Other minor tweaks:
- The login sheet now uses the full domain list stored in `MlemStats` for instance suggestions.
- Fixed a bug where a toast with text of a specific length would be truncated.

Partially completes #1074

### CODE REVIEW

File: Mlem.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -1941,8 +1961,8 @@
 			isa = XCRemoteSwiftPackageReference;
 			repositoryURL = "https://github.com/mlemgroup/MlemMiddleware.git";
 			requirement = {
-				kind = upToNextMinorVersion;
-				minimumVersion = 0.15.0;
+				branch = "sjmarf/block-list";
+				kind = branch;
```
Comment (EricBAndrews) [inline Mlem.xcodeproj/project.pbxproj]: Reminder

File: Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+    
+    var body: some View {
+        VStack(spacing: 16) {
+            box {
+                HStack {
+                    Label(instance.created.dateString, systemImage: Icons.cakeDay)
+                    Text("•")
+                    Label(instance.created.getRelativeTime(unitsStyle: .abbreviated), systemImage: Icons.time)
+                }
+                .foregroundStyle(.secondary)
```
Comment (EricBAndrews) [inline Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift]:
```suggestion
                .foregroundStyle(palette.secondary)
```

File: Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+                .font(.footnote)
+            }
+            HStack(spacing: 16) {
+                box {
+                    Text("Users")
+                        .foregroundStyle(.secondary)
```
Comment (EricBAndrews) [inline Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift]:
```suggestion
                        .foregroundStyle(palette.secondary)
```

File: Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+                    Text("\((instance.userCount_ ?? 0).abbreviated)")
+                        .font(.title)
+                        .fontWeight(.semibold)
+                }
+                
+                box {
+                    Text("Communities")
+                        .foregroundStyle(.secondary)
```
Comment (EricBAndrews) [inline Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift]:
```suggestion
                        .foregroundStyle(palette.secondary)
```

File: Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+                    Text("\((instance.communityCount_ ?? 0).abbreviated)")
+                        .font(.title)
+                        .fontWeight(.semibold)
+                        .foregroundStyle(.green)
```
Comment (EricBAndrews) [inline Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift]: This should be part of `palette`--we have several situations where we want an accent color that isn't attached to any particular concept, we can capture them in `palette` as `accent1`, `accent2`, etc.

File: Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+            .frame(maxWidth: .infinity)
+            
+            HStack(spacing: 16) {
+                box {
+                    Text("Posts")
+                        .foregroundStyle(.secondary)
+                    Text("\((instance.postCount_ ?? 0).abbreviated)")
+                        .font(.title)
+                        .fontWeight(.semibold)
+                        .foregroundStyle(.pink)
```
Comment (EricBAndrews) [inline Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift]: palette accent

File: Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+                }
+                
+                box {
+                    Text("Comments")
+                        .foregroundStyle(.secondary)
+                    Text("\((instance.commentCount_ ?? 0).abbreviated)")
+                        .font(.title)
+                        .fontWeight(.semibold)
+                        .foregroundStyle(.orange)
```
Comment (EricBAndrews) [inline Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift]: palette accent

File: Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+                        .fontWeight(.semibold)
+                        .foregroundStyle(.green)
+                }
+            }
+            .frame(maxWidth: .infinity)
+            
+            HStack(spacing: 16) {
+                box {
+                    Text("Posts")
+                        .foregroundStyle(.secondary)
```
Comment (EricBAndrews) [inline Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift]:
```suggestion
                        .foregroundStyle(palette.secondary)
```

File: Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+                    Text("\((instance.postCount_ ?? 0).abbreviated)")
+                        .font(.title)
+                        .fontWeight(.semibold)
+                        .foregroundStyle(.pink)
+                }
+                
+                box {
+                    Text("Comments")
+                        .foregroundStyle(.secondary)
```
Comment (EricBAndrews) [inline Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift]:
```suggestion
                        .foregroundStyle(palette.secondary)
```

File: Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+                        .fontWeight(.semibold)
+                        .foregroundStyle(.orange)
+                }
+            }
+            .frame(maxWidth: .infinity)
+            
+            if let activeUserCount = instance.activeUserCount_ {
+                box(spacing: 8) {
+                    Text("Active Users")
+                        .foregroundStyle(.secondary)
```
Comment (EricBAndrews) [inline Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift]:
```suggestion
                        .foregroundStyle(palette.secondary)
```

File: Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+            .frame(maxWidth: .infinity)
+            .background(palette.secondaryGroupedBackground)
+            .cornerRadius(AppConstants.largeItemCornerRadius)
+            
+            VStack(alignment: .leading, spacing: 0) {
+                settingRow(
+                    "Registration",
+                    systemImage: Icons.person,
+                    value: instance.registrationMode_?.label ?? "Closed",
+                    color: instance.registrationMode_?.color ?? .red
```
Comment (EricBAndrews) [inline Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift]:
```suggestion
                    color: instance.registrationMode_?.color ?? palette.negative```

File: Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+                        "Email Verification",
+                        systemImage: Icons.email,
+                        value: instance.emailVerificationRequired_ ?? false
+                    )
+                    Divider()
+                    settingRow(
+                        "Captcha",
+                        systemImage: Icons.photo,
+                        value: captchaLabel,
+                        color: instance.captchaDifficulty_ == nil ? .red : .green
```
Comment (EricBAndrews) [inline Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift]:
```suggestion
                        color: instance.captchaDifficulty_ == nil ? palette.negative: palette.positive
```

File: Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+                    "Slur Filter",
+                    systemImage: Icons.filterFill,
+                    value: instance.slurFilterRegex_ != nil
+                )
+                if let regex = instance.slurFilterRegex_ {
+                    Divider()
+                    VStack(alignment: .leading, spacing: 2) {
+                        if showingSlurRegex {
+                            Text(regex)
+                                .foregroundStyle(.secondary)
```
Comment (EricBAndrews) [inline Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift]:
```suggestion
                                .foregroundStyle(palette.secondary)
```

File: Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift [hunk 14]
```diff
@@ -0,0 +1,296 @@
...
+                                .textSelection(.enabled)
+                        } else {
+                            Text("Tap to show slur filter regex.")
+                            Label(
+                                "This probably contains foul language.",
+                                systemImage: Icons.warning
+                            )
+                            .foregroundStyle(.orange)
```
Comment (EricBAndrews) [inline Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift]: We should add this to palette as `caution`

File: Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift [hunk 15]
```diff
@@ -0,0 +1,296 @@
...
+                )
+                Divider()
+                settingRow(
+                    "Reports Email Admins",
+                    systemImage: Icons.moderationReport,
+                    value: instance.emailReportsToAdmins_ ?? false
+                )
+            }
+            .frame(maxWidth: .infinity)
+            .background(Color(uiColor: .secondarySystemGroupedBackground))
```
Comment (EricBAndrews) [inline Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift]:
```suggestion
            .background(palette.secondaryGroupedBackground)
```

File: Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift [hunk 16]
```diff
@@ -0,0 +1,296 @@
...
+        return "No"
+    }
+    
+    @ViewBuilder func box(spacing: CGFloat = 5, @ViewBuilder content: () -> some View) -> some View {
+        VStack(spacing: spacing) {
+            content()
+        }
+        .padding(.vertical, 10)
+        .frame(maxWidth: .infinity)
+        .background(Color(uiColor: .secondarySystemGroupedBackground))
```
Comment (EricBAndrews) [inline Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift]:
```suggestion
        .background(palette.secondaryGroupedBackground)
```

File: Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift [hunk 17]
```diff
@@ -0,0 +1,296 @@
...
+        .cornerRadius(AppConstants.largeItemCornerRadius)
+    }
+    
+    @ViewBuilder func settingRow(
+        _ label: String,
+        systemImage: String,
+        value: String,
+        color: Color = .primary
```
Comment (EricBAndrews) [inline Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift]: Or `Palette.main`, not sure how nicely default parameters play with environment

```suggestion
        color: Color = palette.primary
```

File: Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift [hunk 18]
```diff
@@ -0,0 +1,296 @@
...
+    ) -> some View {
+        HStack {
+            Image(systemName: systemImage)
+                .foregroundStyle(.secondary)
```
Comment (EricBAndrews) [inline Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift]:
```suggestion
                .foregroundStyle(palette.secondary)
```

File: Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift [hunk 19]
```diff
@@ -0,0 +1,296 @@
...
+            Text(value)
+                .foregroundStyle(color)
+        }
+        .padding(12)
+    }
+    
+    @ViewBuilder func settingRow(_ label: String, systemImage: String, value: Bool) -> some View {
+        HStack {
+            Image(systemName: systemImage)
+                .foregroundStyle(.secondary)
```
Comment (EricBAndrews) [inline Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift]:
```suggestion
                .foregroundStyle(palette.secondary)
```

File: Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift [hunk 20]
```diff
@@ -0,0 +1,296 @@
...
+                .frame(width: 30)
+            Text(label)
+            Spacer()
+            Text(value ? "Yes" : "No")
+                .foregroundStyle(value ? .green : .red)
```
Comment (EricBAndrews) [inline Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift]:
```suggestion
                .foregroundStyle(value ? palette.positive: palette.negative)
```

File: Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift [hunk 21]
```diff
@@ -0,0 +1,296 @@
...
+    }
+    
+    @ViewBuilder
+    func activeUserBox(_ label: String, value: Int) -> some View {
+        VStack {
+            Text(value.abbreviated)
+                .font(.title3)
+                .fontWeight(.semibold)
+            Text(label)
+                .foregroundStyle(.secondary)
```
Comment (EricBAndrews) [inline Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift]:
```suggestion
                .foregroundStyle(palette.secondary)
```

File: Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift [hunk 22]
```diff
@@ -0,0 +1,296 @@
...
+    }
+    
+    var color: Color {
+        switch self {
+        case .closed:
+            return .red
+        case .requireApplication:
+            return .orange
+        case .open:
+            return .green
```
Comment (EricBAndrews) [inline Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift]: palette negative/caution/positive

File: Mlem/App/Views/Pages/Instance/InstanceView.swift [hunk 23]
```diff
@@ -0,0 +1,118 @@
...
+    @State var upgradeState: LoadingState = .idle
+    
+    @State var selectedTab: Tab = .about
+    @State var isAtTop: Bool = true
+    
+    // If != nil, blocking is in progress
+    @State var isBlocking: UUID?
+    
+    init(instance: any InstanceStubProviding) {
+        self.instance = instance
```
Comment (EricBAndrews) [inline Mlem/App/Views/Pages/Instance/InstanceView.swift]: Fails build on Xcode 
```suggestion
        self._instance = .init(wrappedValue: instance)
```

File: Mlem/App/Views/Shared/Navigation/NavigationPage.swift [hunk 24]
```diff
@@ -59,7 +64,10 @@ extension NavigationPage {
         case .inbox:
             InboxView()
         case .search:
-            EmptyView()
+            VStack {
+                Link("Test", destination: URL(string: "https://lemmy.dbzer0.com")!)
+                Text("\((AppState.main.firstSession as? UserSession)?.blocks?.instanceCount)")
```
Comment (EricBAndrews) [inline Mlem/App/Views/Shared/Navigation/NavigationPage.swift]: Fails build on Xcode 15

```suggestion
                Text("\((AppState.main.firstSession as? UserSession)?.blocks?.instanceCount ?? 0)")
```

**Review by EricBAndrews — CHANGES_REQUESTED** [review]
Looks good!

Some `palette` conformance and build failures to address, plus design of 'visit' feature (discussion in Slack)

File: Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift (hunk 5 above)
Comment (Sjmarf) [reply Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift]: Should these not be entity-attached - `palette.post`, `palette.comment` etc? We use these same colors for the icons in Settings and in community details. If these were accents, we'd end up with either having two `.orange` accent colors or having the lock color be the same as the comment color with no way to separate them. We'd also have to remember to use `accent2` when referring to comments, etc.

Comment (Sjmarf) [reply Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift]: I suppose we could use numbered accents and then have computed properties of `Palette` for `postAccent`, `commentAccent` etc that return the accent values?

```swift
var postAccent: Color { accent3 }
```

Comment (EricBAndrews) [reply Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift]: That solution sounds perfect to me--we can get very granular semantic colors without needing to redefine forty duplicates in every palette

Comment (Sjmarf) [reply Mlem/App/Views/Pages/Instance/InstanceDetailsView.swift]: Done 👍

**Review by EricBAndrews — APPROVED** [review]
LGTM! 🚀

**Pull request merged**